### PR TITLE
feat(op-node): Fetch receipts directly from `reth`'s DB

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Small experiment that introduces a linked Rust dylib into `op-service` that can read receipts directly from disk using `reth`'s Database bindings. This is introduced as a special `rpckind` within the `op-node`, allowing the `op-node` to optionally remove the need for RPC communication when fetching receipts if it has access to an L1 `reth` datadir.

### TODO
- [x] Set up initial Rust dylib
    - [x] Clean up hacks
    - [x] Gracefully error
- [x] Add Cgo bindings in `op-service` for calling out to the dylib
- [x] Add new `rpckind` flag and `l1.rethdb` flag to the `op-node` and forward this flag down to the receipt fetching job.
- [x] Swap from raw receipts to hydrated receipts returned by the jsonRPC. 

### Running this Version

1. Build the `rethdb-reader` dylib inside of `op-service` with stable `rustc`: `cargo build --release`
2. Build the `op-node` with the `rethdb` buildtag enabled: `go build -tags rethdb -o ./bin/op-node ./cmd/main.go`

### How it Works

The `op-node` now has a new `rpckind`, `reth_db`, as well as a new flag that allows for supplying the DB path. In the `receiptsFetchingJob` in `op-service`, a new alternative method is defined for the `reth_db` rpckind that calls out to the Rust dylib.

This dylib directly accesses the L1 node's database on-disk to fetch the receipts belonging to a blockhash, serializes them in JSON, and passes them back to Go through the `cgo` interface to be deserialized into an array of `geth`'s `Receipt` type.

This feature is also hidden behind the `rethdb` build tag - without the tag enabled, the new fetching method is stubbed and an attempt to link the dylib will not be made.

### Rough Benchmarks

Benchmarks measured from the starting time of the `op-node`, with a completely fresh database, syncing the first `10,000` blocks of `Base Goerli`.

**Setup:**
1. L1: `reth`
2. L2: `op-reth`
3. L2 CL: `op-node`

Standard RPCKind: `1m57s`
RethDB "RPCKind": `1m41s`

### CODE REVIEW

File: op-service/sources/receipts.go [hunk 1]
```diff
@@ -268,6 +270,18 @@ const (
...
 	ErigonGetBlockReceiptsByBlockHash
+	// RethGetBlockReceiptsMDBX is a Reth-specific receipt fetching method. It reads the data directly from reth's database, using their
+	// generic DB abstractions, rather than requesting it from the RPC provider.
+	// Available in:
+	//   - Reth
+	// Method: n/a - does not use RPC.
+	// Params:
+	//   - Reth: string, hex-encoded block hash
+	// Returns:
+	//   - Reth: array of RLP-encoded receipts
```
Comment (clabby) [inline op-service/sources/receipts.go]: This is wrong; need to update

File: op-service/sources/reth_db_stub.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// FetchRethReceipts fetches the receipts for the given block hash...
+func FetchRethReceipts(dbPath string, blockHash *common.Hash) (types.Receipts, error) {
+	panic("unimplemented!")
+}
```
Comment (clabby) [inline op-service/sources/reth_db_stub.go]:
```suggestion
// FetchRethReceipts stub; Not available without `rethdb` build tag.
func FetchRethReceipts(dbPath string, blockHash *common.Hash) (types.Receipts, error) {
	panic("unimplemented! Did you forget to enable the `rethdb` build tag?")
}
```

File: op-node/node/config.go [hunk 3]
```diff
@@ -60,6 +60,9 @@ type Config struct {
 
 	// Cancel to request a premature shutdown of the node itself, e.g. when halting. This may be nil.
 	Cancel context.CancelCauseFunc
+
+	// [OPTIONAL] The reth DB path to read receipts from
+	RethDBPath *string
```
Comment (Inphi) [inline op-node/node/config.go]: nit: empty string is good as defaults in Go.

File: op-service/sources/reth_db.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+	if receiptsResult.error {
+		return nil, fmt.Errorf("Error fetching receipts from Reth Database.")
+	}
+
+	// Convert the returned JSON string to Go string and parse it
+	receiptsJSON := C.GoStringN(receiptsResult.data, C.int(receiptsResult.data_len))
+	var receipts types.Receipts
+	if err := json.Unmarshal([]byte(receiptsJSON), &receipts); err != nil {
+		return nil, err
```
Comment (Inphi) [inline op-service/sources/reth_db.go:61]: if this fails, we leak`receiptsResult.data` memory.

File: op-service/rethdb-reader/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+mod receipts;
+
+/// Read the receipts for a blockhash from the RETH database directly.
+///
+/// # Safety
+/// - All possible nil pointer dereferences are checked, and the function will return a
+///   failing [ReceiptsResult] if any are found.
+#[no_mangle]
+pub unsafe extern "C" fn read_receipts(
```
Comment (Inphi) [inline op-service/rethdb-reader/src/lib.rs]: I'd prefix this symbol and `free_string` with `rethdb_reader_` or something. `read_receipts` is too generic.

**Review by Inphi — COMMENTED** [review]
I haven't looked into the rust code as I'm not familiar with op-reth, but it'll be good to have some unit tests included.

**Review by Inphi — APPROVED** [review]
LGTM! We should follow up on adding tests on another PR.

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #7893** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7893" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7893?utm_source=stack-comment).

File: op-service/rethdb-reader/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+Documentation is available via rustdoc.
+
+```sh
+cargo doc --open
+```
+
+**Linting**
+
+```sh
+cargo +nightly fmt -- && cargo +nightly clippy --all --all-features -- -D warnings
```
Comment (trianglesphere) [inline op-service/rethdb-reader/README.md:27]: Do we need to depend on nightly?

File: op-node/node/config.go [hunk 7]
```diff
@@ -60,6 +60,9 @@ type Config struct {
 
 	// Cancel to request a premature shutdown of the node itself, e.g. when halting. This may be nil.
 	Cancel context.CancelCauseFunc
+
+	// [OPTIONAL] The reth DB path to read receipts from
+	RethDBPath string
```
Comment (trianglesphere) [inline op-node/node/config.go:65]: This is kinda awkward. It feels like it should be in L1EndpointSetup, but I'm also not sure what's in L1EndpointSetup

File: op-service/sources/reth_db_stub.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+package sources
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// FetchRethReceipts stub; Not available without `rethdb` build tag.
+func FetchRethReceipts(dbPath string, blockHash *common.Hash) (types.Receipts, error) {
+	panic("unimplemented! Did you forget to enable the `rethdb` build tag?")
```
Comment (trianglesphere) [inline op-service/sources/reth_db_stub.go:12]: When / where do we set the build tag?

Comment (clabby) [reply op-service/sources/reth_db_stub.go:12]: When compiling the `op-node`, the `rethdb` build tag can be set like so (modification of the make directive, purposefully didn't add a new directive as this feature is experimental):

`env GO111MODULE=on GOOS=$(TARGETOS) GOARCH=$(TARGETARCH) go build -tags rethdb -v $(LDFLAGS) -o ./bin/op-node ./cmd/main.go`

File: op-node/node/config.go (hunk 7 above)
Comment (clabby) [reply op-node/node/config.go:65]: Glad to swap things around - feel free to throw up a ticket and I'll grab it :)

File: op-service/rethdb-reader/README.md (hunk 6 above)
Comment (clabby) [reply op-service/rethdb-reader/README.md:27]: The nightly `cargo-fmt` + `cargo-clippy` is the standard for linting - it's pretty rare that projects will use the stable formatter or clippy. There are many more useful lints etc. in the nightly toolchain, so most people will build / test on stable but lint on nightly.

**Pull request merged**